DOCS-8239-streaming-duke (#1508)
mulesoft/docs-mule-runtime-3: modules/ROOT/pages/streaming-about.adoc

 </sfdc:query>
----

[[disable_repeatable_streaming]]
== Disabling Repeatable Streaming

> You can disable repeatable streaming through the `non-repeatable-stream` and `non-repeatable-iterable` strategies. The strategy to use depends on the type of stream.

Reviewer: Should we clarify what we mean by type of stream? I don't see that in the following text.

Author: type of stream is covered in preceding (stream of objects or bytes)

> Use this option only if you are certain that there is no need to consume the stream several times and only if you need a very tight optimization for performance and resource consumption. 

Reviewer: If you want to emphasize this even more, try list format:
```suggestion
Use this option only if both of the following are true:
* You are certain that there is no need to consume the stream several times.
* You need tight optimization for performance and resource consumption. 
```
You might also consider whether it is more effective to say there is not need to consume several times, or the stream will be consumed only one or a few times.

> When making that decision, consider the following: 

Reviewer: This seems a little out of place, or repetitive of the two items just listed: frequency of consumption and very strict optimization.
Do the following texts belong under one of the two items I am suggesting that we list, as further info on each?

Reviewer: If you keep the texts where they are, do we really need this intro sentence?
```suggestion
```

> * Although disabling repeatable streaming improves performance, the significance of that gain is proportional to the size of the streams. Unless you are working with large streams, it is likely that you can do something else in your app to provide a better optimization.

Reviewer: Yeah, I would put these bullets as text following the suggested previous bullets:

* You are certain that there is no need to consume the stream several times. +
Disabling repeatable streaming means that the stream can only be read once. Keep in mind that your flow might contain components that require consumption of the full stream even if you do not explicitly specify that requirement, such as:
** The Cache component (`<ee:cache>`)
**Some transformations through the Transform Message component (`<ee:transform>`)
**The For Each component (`<foreach>`) when iterating through a JSON array
**Any other component that accesses the stream directly or through an expression

* You need tight optimization for performance and resource consumption. +
Although disabling repeatable streaming improves performance, the significance of that gain is proportional to the size of the streams. Unless you are working with large streams, it is likely that you can do something else in your app to provide a better optimization.    //for example?//

Author: sounds good. I will try this in the future. :)

> * Disabling repeatable streaming means that the stream is only read once. Keep in mind that your flow might contain components that require consumption of the full stream even if you do not explicitly specify that requirement. Examples include the Cache component (`<ee:cache>`), some transformations through the Transform Message component (`<ee:transform>`, the For Each component (`<foreach>`) when iterating through a JSON array, and any other component that is using an expression that accesses the stream. 

Reviewer 2: `means that the stream is only read once` -> `means that the stream can only be read once`

Author: done

> * Disabling repeatable streaming means that the stream is only read once. Keep in mind that your flow might contain components that require consumption of the full stream even if you do not explicitly specify that requirement. Examples include the Cache component (`<ee:cache>`), some transformations through the Transform Message component (`<ee:transform>`, the For Each component (`<foreach>`) when iterating through a JSON array, and any other component that is using an expression that accesses the stream. 

Reviewer 2: `the Transform Message component (<ee:transform>` -> `the Transform Message component (<ee:transform>)`

Author: done

> * Disabling repeatable streaming means that the stream is only read once. Keep in mind that your flow might contain components that require consumption of the full stream even if you do not explicitly specify that requirement. Examples include the Cache component (`<ee:cache>`), some transformations through the Transform Message component (`<ee:transform>`, the For Each component (`<foreach>`) when iterating through a JSON array, and any other component that is using an expression that accesses the stream. 

Reviewer 2: `and any other component that is using an expression that accesses the stream. ` -> `and any other component that accesses the stream directly or through an expression`

> * Disabling repeatable streaming means that the stream is only read once. Keep in mind that your flow might contain components that require consumption of the full stream even if you do not explicitly specify that requirement. Examples include the Cache component (`<ee:cache>`), some transformations through the Transform Message component (`<ee:transform>`, the For Each component (`<foreach>`) when iterating through a JSON array, and any other component that is using an expression that accesses the stream. 

Author: ```suggestion
* Disabling repeatable streaming means that the stream can only be read once. Keep in mind that your flow might contain components that require consumption of the full stream even if you do not explicitly specify that requirement. Examples include the Cache component (`<ee:cache>`), some transformations through the Transform Message component (`<ee:transform>`), the For Each component (`<foreach>`) when iterating through a JSON array, and any other component that accesses the stream directly or through an expression. 
```

 </sfdc:query>
----

[[non_repeatable_iterable]]
=== Non-Repeatable Iterable

> This non-repeatable strategy disables repeatable streams, which enables you to write a stream of output objects only once. Use this strategy when your use case does not require the extra memory or performance overhead that comes with repeatable streaming. Because the stream is not saved to memory, the non-repeatable strategy performs better than the other strategies.

Reviewer 2: This entire paragraph is wrong, but so is the non-repeatable-stream one. I would collapse both sections into a unique one about disable repeatable streaming that reads like this:

```
Repeatable streaming can be disabled through the non-repeatable-stream or non-repeatable-iterable strategies, depending on the type of stream.

This option should only be used when you are certain that there's no need to consume the stream several times and you need a very tight optimization for performance and resource consumption. 

When making that decision, you should consider the following: 

* Although disabling repeatable streaming will improve performance, the significance of that gain will be proportional to the size of the streams themselves. Unless you're dealing with large streams, chances are that there are others things you can improve in your app that will provide a better optimization.
* Disabling repeatable streaming means that the stream can only be read once. When trying to determine if that's the case, keep in mind that your flow may contain components that require to consume the full stream even if you don't explicitly specify that. Examples include the <ee:cache> component, <ee:transform> (depending on the actual transformation being used), <foreach> (when trying to iterate through a Json array), or any other component using an expression which accesses the stream. 
```

Author: added and copyedited per style guidelines
